test/framework/src/main/java/org/elasticsearch/bootstrap/TestScopeResolver.java

```diff
@@ -24,14 +24,24 @@
 
 import static org.elasticsearch.entitlement.runtime.policy.PolicyManager.ALL_UNNAMED;
 import static org.elasticsearch.entitlement.runtime.policy.PolicyManager.ComponentKind.PLUGIN;
+import static org.elasticsearch.entitlement.runtime.policy.PolicyManager.ComponentKind.SERVER;
 
 public record TestScopeResolver(Map<String, PolicyManager.PolicyScope> scopeMap) {
 
     private static final Logger logger = LogManager.getLogger(TestScopeResolver.class);
 
     PolicyManager.PolicyScope getScope(Class<?> callerClass) {
         var callerCodeSource = callerClass.getProtectionDomain().getCodeSource();
-        assert callerCodeSource != null;
+        if (callerCodeSource == null) {
+            // This case happens for JDK modules. Usually those are trivially allowed, but some are excluded,
+            // and those end up here.
+            // We have no test build info for those modules, so for now, let's just guess.
+            if (callerClass.getPackageName().equals("sun.java2d")) {
```

Comment:
I understand why this needs to be a special case and reflect what happens in production, but I do wonder if we can come up with a more generic solution, so we don't have to keep adding things here.

What if we add a step/utility to our test framework, that inspects each module in `MODULES_EXCLUDED_FROM_SYSTEM_MODULES` and builds a map `packageName -> moduleName` for each package in each module, and / or `moduleName (+ component, always server) -> PolicyScope`? That would generalize this code and it would work with any package from any of these "JDK modules which are not in the system layer" modules.

It should not be too difficult, but maybe it's better to handle this in a follow-up. Wdyt?

Reply:
Sounds like a dynamic version of Test Build Info? That makes sense to me. We could also do it statically, actually, using the existing Gradle task, assuming we're able to know what JDK we're using for tests.